File: 0078-subsets/0078-subsets.java
class Solution {
    private List<List<Integer>> res = new ArrayList<>();
    private int[] nums;

    private void generateSubarray(List<Integer> path, int idx){
        if(idx == nums.length){
            res.add(path);
            return;
        }

        List<Integer> temp = new ArrayList<>(path);
        temp.add(nums[idx]);

        generateSubarray(path, idx+1);
        generateSubarray(temp, idx+1);
    }

    public List<List<Integer>> subsets(int[] nums) {
        this.nums = nums;
        List<Integer> path = new ArrayList<>();
        generateSubarray(path, 0);
        return res;
    }
}

File: 0078-subsets/0078-subsets.py
class Solution(object):
    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        n = len(nums)
        res = []

        def backtrack(idx, path):
            if idx == n:
                res.append(path)
                return

            #pick
            backtrack(idx+1, path + [nums[idx]])

            #not pick
            backtrack(idx+1, path)
        
        backtrack(0, [])
        return res
            